_episodes/08-edge-detection.md

```diff
@@ -49,430 +45,376 @@ black and and one white:
+user interface and sliders.
 
 Here is a Python program that shows how to apply Canny edge detection, and how
-to add trackbars to the user interface. There are four parts to this program, 
-making it a bit (but only a *bit*) more complicated that the programs we have 
-looked at so far. The added complexity comes from three *functions* we have 
-written. From top to bottom, the parts are:
-
-* The `cannyEdge()` function, 
-* The `adjustMin()` function, 
-* the `adjustMax()` function, and
+to add sliders to the user interface. There are four parts to this program,
+making it a bit (but only a *bit*) more complicated than the programs we have
+looked at so far.
+The added complexity comes from setting up the sliders for the parameters that were previously read from the command line:
+In particular, we have added
+
+* The `canny()` filter function that returns an edge image,
+* The `cannyPlugin` plugin object, to which we add
+* The sliders for *sigma*, and *low* and *high threshold* values, and
 * The main program, i.e., the code that is executed when the program runs.
 
-We will look at the main program part first, and then return to the three 
-functions. The first several lines of the main program are easily recognizable
-at this point: saving the command-line argument, reading the image in 
-grayscale, and creating a window. 
+We will look at the main program part first, and then return writing the Plugin.
+The first several lines of the main program are easily recognizable
+at this point: saving the command-line argument, reading the image in
+grayscale, and creating a window.
 
 ~~~
-'''
- * Main program begins here. 
-'''
-# read command-line filename argument
-filename = sys.argv[1]
+"""
+ * Python script to demonstrate Canny edge detection
+ * with trackbars to adjust the thresholds.
+ *
+ * usage: python CannyTrack.py <filename>
+"""
+import skimage
+import skimage.feature
+import skimage.viewer
+import sys
 
-# load original image as grayscale
-image = cv2.imread(filename = filename, flags = cv2.IMREAD_GRAYSCALE)
 
-# set up display window with trackbars for minimum and maximum threshold
-# values
-cv2.namedWindow(winname = "edges", flags = cv2.WINDOW_NORMAL)
+filename = sys.argv[1]
+image = skimage.io.imread(fname=filename, as_gray=True)
+viewer = skimage.viewer.ImageViewer(image)
 ~~~
 {: .python}
 
-Then, the program creates two variables to
-hold first guesses for the low and high threshold values, `minT` and `maxT`. 
+The `skimage.viewer.plugins.Plugin` class is designed to manipulate images.
+It takes an `image_filter` argument that should be a function.
+This function should produce a new image as an output which then will be automatically displayed in the image viewer.
+With this in mind, we write a function to perform Canny filtering, with an image as the first parameter, followed by sigma, and low and high threshold values
 
 ~~~
-minT = 30
-maxT = 150
+def canny(image, sigma, low_threshold, high_threshold):
```

Comment:
Why do we wrap `canny` in a new function here?
Wouldn't it be less code and also easier if we just passed `skimage.feature.canny` to the `Plugin` below?

Reply:
I thought it would make it clearer, that you can write _any_ function and push it to a plugin, as long as it takes and produces an image, but yeah, point taken.

Reply:
Ok, I see your point.
I would still prefer to not writing a separate function. 
It just makes the code cleaner here and I think it's less mental overhead.